drivers: ethernet: xlnx_gem: fixes for UltraScale operation

ibirnbaum opened the pull request:
Fixes the issue of the transmit logic locking up whenever the driver is running on actual UltraScale+ hardware, where an additional set of RX/TX buffer descriptors can be specified via an additional set of RXQ base / TXQ base registers, whose primary purpose it seems to be to make DMA access 64-bit compatible (a high address word register also exists for each of the additional RXQ base / TXQ base registers). However, if at least the alternate TXQ base register is left in its power-on state and buffer DMA initialization is only performed identically to how it's being performed on the Zynq-7000, the GEM's TX path locks up when attempting to transmit a packet. Further info on the symptoms of this issue can now be found in the driver's buffer configuration function.

This fix solves the issue in the most simple way possible, by setting up an extra DMA area for the 2nd set of RXQ base / TXQ base registers. This DMA area effectively only consists of a single TX buffer descriptor, which will never indicate pending TX data to the controller, also, its control word contains the 'wrap' bit so that the 2nd TXQ base register will never point to anything else than this dummy TX buffer descriptor. However, placing this single descriptor in a new type of DMA area struct retains the option to expand the capabilities of the 2nd DMA area in case full 64-bit support might eventually be introduced for the UltraScale's A53 cores.

This commit also fixes the UltraScale+-specific issue hinted at in the dt-bindings commit: the driver performs a validity check in the form of an __ASSERT macro regarding the MDC clock divider. This check has so far been specific to whether the current system is based on the Zynq-7000 or the UltraScale+, with different maximum dividers for both SoC families. According to the most recent documentation, treating the two SoC families differently is not required when it comes to this particular validity check.

Another minor issue fixed is that https://github.com/zephyrproject-rtos/zephyr/pull/41918 apparently didn't automatically fix device config/data usage if the respective function was #ifdef'd. Therefore, compilation of eth_xlnx_gem_stats failed whenenver CONFIG_NET_STATISTICS_ETHERNET was set.

As a side effect of refactoring, register bits which so far were specified as full 32-bit hex words, now use the BIT(x) notation.

ibirnbaum committed:
dt-bindings: fix MDC clock divider bindings for Xilinx GEM

The original implementation of the GEM's device tree binding
implied that:
1) on the UltraScale+, the MDC clock divider is applied to
the LPD_LSBUS_CLK. According to the most recent documentation,
this is not the case, instead, the MDC divider is applied to
the IOU_SWITCH_CLK.
2) any MDC divider greater than 32 is reserved to the Zynq-7000
(in the driver itself, accessibility of the larger dividers was
also #ifdef'd), as the Zynq's MDC clock source, the cpu_1x
clock, can have significantly higher frequencies than the
UltraScale's LPD_LSBUS clock.

The respective documentation in the device tree binding header
file is hereby fixed.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

yashi commented:
> Fixes the issue of the transmit logic locking up whenever the driver is running on actual UltraScale+ hardware, where an additional set of RX/TX buffer descriptors can be specified via an additional set of RXQ base / TXQ base registers [...]

Would you mind which issue you are referring to?  Is it [52019](https://support.xilinx.com/s/article/52019)?

yashi commented:
Hi @ibirnbaum,

Thank you for your detailed explanation.  Now I see what's going on.  I was gonna test it with my Zynq-7000 but I'm sure I can't.

> If the UltraScale revision is detected, the 2nd TX queue is set up as a full-fledged TX queue, including making use of the corresponding 2nd set of interrupt enable/disable/status registers and handling them in the ISR. However, as observed by anyone running into this issue, the RX buffer descriptor queue is present only once, and is referred to by the original RXQBASE register which can be found in both the Zynq and the UltraScale revision of the controller.

It seems all the change is squashed into [this commit](https://github.com/Xilinx/embeddedsw/commit/54cf0f4cfaa2b287b0096e10aaa5b4d631b4b1d7), which doesn't help. :face_exhaling: 

On to the commits.
It might help us to maintain the change if the commit "drivers: ethernet: xlnx_gem: fixes for UltraScale operation" is specific to the problem you described.  Would it be too much to ask splitting minor changes out to its own?

ibirnbaum commented:
@yashi
> On to the commits. It might help us to maintain the change if the commit "drivers: ethernet: xlnx_gem: fixes for UltraScale operation" is specific to the problem you described. Would it be too much to ask splitting minor changes out to its own?

Would you prefer those other minor changes as separate commits within this PR, or would you rather prefer a separate PR?

yashi commented:
@ibirnbaum

In this PR should be OK.

ibirnbaum committed:
drivers: ethernet: xlnx_gem: DMA fix for UltraScale+

Prevent the TX DMA from locking up on actual hardware UltraScale+
targets by providing a 2nd (dummy) TX buffer descriptor queue
pointed to by the transmit_q1_ptr register. The data structures
for the 2nd TX BD queue currently contain the bare minimum of
data, but could eventually be extended into a fully functional
2nd set of RX/TX DMAs - although the 2nd set of interrupt status/
enable/clear registers are lacking most error handling-related
bits.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: remove SoC dependency for MDC clock

The assumption that certain MDC clock dividers are available
in the Zynq-7000 only has turned out wrong, also, the MDC clock
is driven by the IOU_SWITCH_CLK instead of the LPD_LSBUS_CLK
in the UltraScale+. This commit fixes the MDC divider validity
check by making it universal and makes dividers > 48 available
to the UltraScale+. Documentation of the source clock is updated
as well.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: fix dev data access missed by #41918

Fixes a compiler error in eth_xlnx_gem_stats() caused by the
automated replacement of device run-time data pointer accesses
by #41918. The device run-time data access in eth_xlnx_gem_stats()
was likely missed as this function is not usually compiled - its
existance depends on CONFIG_NET_STATISTICS_ETHERNET being set.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: fix missing CLKACT bit

Although most likely activated by the PS7Init code anyways,
explicitly set the CLKACT bit whenever the clock configuration
is adjusted, either at init time, or whenever a link speed
chage occurrs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: adjust IRQ sources

Adjust which interrupt sources are enabled at run-time.
Always clear the entire range of interrupt source bits when
initializing or stopping the device.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: advance RX BD when SOF is missing

Advance the RX buffer descriptor queue to the next entry whenever
a missing start-of-frame (SOF) bit is encountered. If this is case,
until now, the RX logic would keep considering the broken RX BD
as current until the RX BD queue wraps around and overwrites the
RX BD at the current position.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: code formatting, additional defines

- Updates of code comments
- change 32-bit notation for register bits to use of BIT(x)
- bit masks for supported interrupt sources are now explicit
  ORs of the individual bits instead of 32-bit binary notation,
  improving readability
- indentation fixes
- added some register bits which are available both on the
  Zynq-7000 and the UltraScale+ which until now had not been
  specified.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum commented:
> @ibirnbaum
> 
> In this PR should be OK.

Hi @yashi, I've split up what was originally the single driver-related commit into individual ones for each topic addressed.

yashi commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -1449,12 +1449,26 @@ static void eth_xlnx_gem_handle_rx_pending(const struct device *dev)
 		reg_val = sys_read32(reg_ctrl);
```
Maybe we should use another variable for reg_ctrl, so that reg_val can be used in the if block below?

ibirnbaum replied:
The reg_val variable is used twice sequentially in order to evaluate the contents of two different memory locations. First, the /address/ word of the current descriptor in the RX buffer descriptor ring is read and checked for the presence of the 'Used' bit. Afterwards, reg_val is re-used as storage for the contents of the current descriptor's /control/ word, which /should/ contain the Start-Of-Frame bit. So line 1463 isn't a re-read, but a read of another memory location compared to line 1449, where the local variable the data is being read to is being re-used.

yashi replied:
Yeah, I know.  What I meant was you read the location of `reg_addr` twice. 1st at line 1441 and 2nd at 1465 because you overwrite the variable with `reg_val = sys_read32(reg_ctrl)` at 1449.

But, I assume it's a minor issue and I approve this PR anyway.

yashi commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -1449,12 +1449,26 @@ static void eth_xlnx_gem_handle_rx_pending(const struct device *dev)
 				dev->name, first_bd_idx);
-			break;
+
+			reg_val = sys_read32(reg_addr);
```
No need to re-read the register if we use a different variable.  See above.

yashi reviewed (commented):
Hi @ibirnbaum,

Thanks for the update! It basically looks good to me.  

I've added some questions to the patches.  And a few comments on some of the commit logs. 

### drivers: ethernet: xlnx_gem: DMA fix for UltraScale+

> but could eventually be extended into a fully functional
> 2nd set of RX/TX DMAs - although the 2nd set of interrupt status/
> enable/clear registers are lacking most error handling-related
> bits.

Not sure what you are trying to convey.  I assume that what you are trying to say with this patch is to fix the lockup and doesn't prevent anyone from utilizing the 2nd set of DMA buffers.

It'd be nice if you can copy & past some text from https://github.com/zephyrproject-rtos/zephyr/pull/45556#issuecomment-1124800521 into the commit log.  So that it's more clear what the patch is focusing on.

### drivers: ethernet: xlnx_gem: generic refactoring

I'd say that "drivers: ethernet: xlnx_gem: generic refactoring" is not refactoring but non-functional cosmetic changes.  Refactoring, at least to me, is more like structural change in the code.

BTW, I assume that the doc changes aren't relate to any patch in this series, right?  If it does, it should move to the corresponding patch.  But I'm assuming that you have added missing doc from the past changes.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -511,11 +518,28 @@ struct eth_xlnx_dma_area_gem##port {\
+#if defined(CONFIG_SOC_XILINX_ZYNQMP)
+#define ETH_XLNX_GEM_DMA_AREA_Q1_DECL(port) \
+struct eth_xlnx_dma_q1_area_gem##port {\
+	struct eth_xlnx_gem_bd tx_bd[1];\
```
Hi @yashi,
sorry for the long reply delay.

> Not sure what you are trying to convey. I assume that what you are trying to say with this patch is to fix the lockup and doesn't prevent anyone from utilizing the 2nd set of DMA buffers.

That's correct, the immediate purpose is to fix the lockup by adding a single, always idle, TX buffer descriptor but no RX buffer descriptor, as this direction is not affected by the problem. Yet, the 2nd set of RX/TX buffer descriptors /could/ be made fully functional at a later time.

At the time being, from my point of view it's favourable to stick with the original set of RX/TX BD descriptor rings for actual reception and transmission of packets, pointed to the by rx_qbar (@ GEM base + 0x18) and tx_qbar (@ GEM base + 0x1C) registers. This way, the effective RX/TX-related code stays the same for the Zynq-7000 (which doesn't have the 2nd set of registers introduced with the UltraScale), and the UltraScale fix is just this small addition on top which adds this one dummy TX descriptor and lets the 2nd TX queue base address register point to it.

/However/, this would have to change if someone started using the GEM not from the context of the UltraScale's 32-bit R5 cores, but from the 64-bit A53 cores and wanted to place the respective GEM's DMA area in a memory area located above 4GB - it's only the 2nd, UltraScale-only set of registers with their additional high address word registers which enables the placement of RX/TX descriptor rings in memory above 4GB. In that case, the accordingly enlarged buffer descriptor format with two address words would have to be supported. If 64-bit support was to be added, there'd either be a significant amount of relatively similar code #ifdef'd to the respective SoC family, or maybe the driver sources should be split into two variants for improved readability. But as the A53 cores aren't yet relevant to my work, I haven't experimented in this direction yet.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -511,11 +518,28 @@ struct eth_xlnx_dma_area_gem##port {\
+#if defined(CONFIG_SOC_XILINX_ZYNQMP)
+#define ETH_XLNX_GEM_DMA_AREA_Q1_INST(port) \
+static struct eth_xlnx_dma_q1_area_gem##port eth_xlnx_gem##port##_q1_dma_area\
+	__ocm_bss_section __aligned(8);
```
This attribute has turned out as a mistake after another look at the Zynq's Technical Reference Manual.

The entire DMA area of each GEM is always 4k-aligned so that the properties of the respective MMU page table entry can be modified without affecting any non-related data (at least in the Zynq, whose A9 cores have a MMU). With the simplest RX/TX buffer descriptor format which is currently the only one supported by the driver, each descriptor is 8 bytes in size, so the top address of the descriptor rings, which also equals the bottom address of the RX packet data buffers which are next in the DMA area struct, is always 8-byte aligned.

Even if an odd number is specified in the device tree for the rx-buffer-size and tx-buffer-size properties, each RX/TX packet data buffer will always be padded to a 4 byte boundary (that's what the ETH_XLNX_BUFFER_ALIGNMENT define is for). So the top address of the TX packet data buffers, which are last in the DMA area struct, is always at least 4-byte aligned.

The modified ETH_XLNX_GEM_INITIALIZE macro now first places the 4k-aligned, actually useful DMA area in the OCM memory, directly followed by the dummy DMA area which contains only the one TX descriptor. The __aligned(8) attribute should have enforced placement of the dummy DMA area at an 8 byte-aligned address in case the top address of the actual DMA area was only 4 byte-aligned. I did believe that the TRM stated that the base address of each buffer descriptor ring must be 8-byte aligned, however, the documentation of the rx_qbar and tx_qbar registers in the Zynq's TRM states "The descriptors should be aligned at 32-bit boundaries". Therefore, this attribute is obsolete, I'll remove it.

ibirnbaum commented:
> I'd say that "drivers: ethernet: xlnx_gem: generic refactoring" is not refactoring but non-functional cosmetic changes. Refactoring, at least to me, is more like structural change in the code.
> 
> BTW, I assume that the doc changes aren't relate to any patch in this series, right? If it does, it should move to the corresponding patch. But I'm assuming that you have added missing doc from the past changes.

I'll re-word the commit to avoid confusion.

At least the comment in the file's header regarding the DMA area placement should have been changed with #42036, but I missed that one. The added register bits in the header file are those where I wasn't sure initially if Zynq and UltraScale both have them. But as I'm now sure that both revisions of the controller have them, I've added them along the way, although they aren't used for anything yet. The comments in the "DMA fix for UltraScale+" commit all refer to the matter at hand.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -385,8 +385,8 @@
-#define ETH_XLNX_GEM_IXR_ALL_MASK			0x03FC7FFE
-#define ETH_XLNX_GEM_IXR_ERRORS_MASK			0x00000C60
+#define ETH_XLNX_GEM_IXR_ALL_RELEVANT_MASK		0x00000CF2
+#define ETH_XLNX_GEM_IXR_ERRORS_MASK			0x00000C70
```
I looked into the question regarding the TX underrun bit again, I most likely didn't include it because the UltraScale has it but the Zynq doesn't. So this bit will be a SoC-#ifdef'd inclusion into the interrupt source bit mask.

yashi reviewed (dismissed):
LGTM.

stephanosio commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -321,7 +316,7 @@ static void eth_xlnx_gem_isr(const struct device *dev)
 	 * this ISR is being processed.
 	 */
-	sys_write32((0xFFFFFFFF & ~(ETH_XLNX_GEM_IXR_FRAME_RX_BIT |
+	sys_write32((~0 & ~(ETH_XLNX_GEM_IXR_FRAME_RX_BIT |
```
why is this 32-bit masking necessary? it already calls `sys_write32`, which cannot write more than 32-bits anyway.

also if any, it should be:
```suggestion
	sys_write32((~0U & ~(ETH_XLNX_GEM_IXR_FRAME_RX_BIT |
```

stephanosio commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -529,22 +524,22 @@ static int eth_xlnx_gem_start_device(const struct device *dev)
 	/* Clear RX & TX status registers */
-	sys_write32(0xFFFFFFFF, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
-	sys_write32(0xFFFFFFFF, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
+	sys_write32(~0, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
+	sys_write32(~0, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
```
same deal

```suggestion
	sys_write32(~0U, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
	sys_write32(~0U, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
```

but, IMHO, the initial `0xffffffff` seems to be more straight-forward to me and this change seems unnecessary.

ibirnbaum replied:
Yeah, that change was probably unnecessary. No real benefit regarding formatting / readability, no technical benefit. I've reverted them.

stephanosio commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -588,14 +583,14 @@ static int eth_xlnx_gem_stop_device(const struct device *dev)
-	sys_write32(0xFFFFFFFF, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
-	sys_write32(0xFFFFFFFF, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
+	sys_write32(~0, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
+	sys_write32(~0, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
```
ditto

stephanosio commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -1510,7 +1556,7 @@ static void eth_xlnx_gem_handle_rx_pending(const struct device *dev)
 	}
 
 	/* Clear the RX status register */
-	sys_write32(0xFFFFFFFF, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
+	sys_write32(~0, dev_conf->base_addr + ETH_XLNX_GEM_RXSR_OFFSET);
```
ditto

stephanosio commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -1625,7 +1671,7 @@ static void eth_xlnx_gem_handle_tx_done(const struct device *dev)
 	}
 
 	/* Clear the TX status register */
-	sys_write32(0xFFFFFFFF, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
+	sys_write32(~0, dev_conf->base_addr + ETH_XLNX_GEM_TXSR_OFFSET);
```
ditto

ibirnbaum committed:
Merge branch 'main' into gem_ultrascale_fix

yashi commented:
@ibirnbaum, we don't push merge commit in a PR.  If you need, just rebase it.  Maintainers will merge it for you.

ibirnbaum committed:
Revert "Merge branch 'main' into gem_ultrascale_fix"

This reverts commit c2f9aae467c7497f038b8afc583595a479050958, reversing
changes made to cada8049f75c9f0cd51ec3f481a2e8bc1f91a033.

ibirnbaum commented:
Rebase gone wrong as usual.
